pkg/lightning/backend/tidb/tidb.go

```diff
@@ -411,27 +492,50 @@ func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, colu
 		insertStmt.WriteByte(')')
 	}
 	insertStmt.WriteString(" VALUES")
+	return &insertStmt
+}
 
-	// Note: we are not going to do interpolation (prepared statements) to avoid
-	// complication arise from data length overflow of BIT and BINARY columns
+// Only used for TestWriteRowsErrorSkip
+var mockErrorCount = 0
 
-	for i, row := range rows {
-		if i != 0 {
-			insertStmt.WriteByte(',')
+func (be *tidbBackend) execStmts(ctx context.Context, stmtTasks []stmtTask, rows tidbRows) (int, error) {
+	// Retry will be done externally, so we're not going to retry here.
+	tx, err := be.db.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		if !common.IsContextCanceledError(err) {
+			log.L().Error("begin a transaction failed before executing statement", zap.Array("rows", rows), zap.Error(err))
 		}
-		insertStmt.WriteString(string(row))
+		return 0, errors.Trace(err)
 	}
+	for _, stmtTask := range stmtTasks {
+		stmt := stmtTask.stmt
+		_, err = tx.ExecContext(ctx, stmt)
 
-	// Retry will be done externally, so we're not going to retry here.
-	_, err := be.db.ExecContext(ctx, insertStmt.String())
-	if err != nil && !common.IsContextCanceledError(err) {
-		log.L().Error("execute statement failed", zap.String("stmt", redact.String(insertStmt.String())),
-			zap.Array("rows", rows), zap.Error(err))
+		failpoint.Inject("mockNonRetryableError", func() {
+			if mockErrorCount == 0 || mockErrorCount == 2 || mockErrorCount == 5 {
+				err = stderrors.New("mock non-retryable error")
+			}
+			mockErrorCount++
+		})
+
+		if err != nil {
+			if !common.IsContextCanceledError(err) {
+				log.L().Error("execute statement failed inside the transaction", zap.String("stmt", redact.String(stmt)),
+					zap.Array("rows", rows), zap.Error(err))
+			}
+			if rollbackErr := tx.Rollback(); rollbackErr != nil {
+				panic(fmt.Sprintf("execute statement failed: %v, unable to rollback: %v\n", err, rollbackErr))
+			}
+			return stmtTask.rowIdx, errors.Trace(err)
```

Comment:
If we return an error, the error logs in this method will be record many times. I think we shall collect the fail rows, and record them in error log in another place.